dockerfiles/cloudbuild.dhall

```diff
@@ -0,0 +1,266 @@
+          { service : Text
+          , dockerfilePaths : List Text
+          , dockerContext : Optional Text
+          , timeout : Optional Text
+          }
+      , default = { dockerContext = None Text, timeout = Some "10800s" }
+      }
+
+let optionalBuildArg
+    : Text → Optional Text → List Text
+    = λ(n : Text) →
+      λ(o : Optional Text) →
+        merge
+          { Some = λ(a : Text) → [ "--build-arg", n ++ "=" ++ a ]
+          , None = [] : List Text
+          }
+          o
+
+let dockerfilePathsArgs =
+      λ(dockerfilePaths : List Text) →
+        List/concatMap Text Text (λ(f : Text) → [ "-f", f ]) dockerfilePaths
+
+let mkArgs
+    : Text → DockerfileDescription.Type → ServiceDescription.Type → List Text
+    = λ(tag : Text) →
+      λ(desc : DockerfileDescription.Type) →
+      λ(serviceDesc : ServiceDescription.Type) →
+          [ "build"
+          , "-t"
+          , tag
+          , "--cache-from"
+          , "gcr.io/\${PROJECT_ID}/${desc.service}"
+          ]
+        # optionalBuildArg "image" (debInfo_ serviceDesc.debCodename).image
+        # optionalBuildArg "MINA_REPO" serviceDesc.repo
+        # optionalBuildArg "network" serviceDesc.network
+        # optionalBuildArg "MINA_BRANCH" serviceDesc.branch
+        # optionalBuildArg
+            "deb_codename"
+            (debInfo_ serviceDesc.debCodename).debCodename
+        # optionalBuildArg "deb_release" serviceDesc.debRelease
+        # optionalBuildArg "deb_version" serviceDesc.debVersion
+        # serviceDesc.extraArgs
+        # merge
+            { Some =
+                λ(ctx : Text) →
+                  dockerfilePathsArgs desc.dockerfilePaths # [ ctx ]
+            , None = [ "-" ]
+            }
+            desc.dockerContext
+
+let mkScript
+    : Text → DockerfileDescription.Type → ServiceDescription.Type → Text
+    = λ(tag : Text) →
+      λ(desc : DockerfileDescription.Type) →
+      λ(serviceDesc : ServiceDescription.Type) →
+            "docker pull gcr.io/\${PROJECT_ID}/${desc.service} || true; "
+        ++  merge
+              { Some = λ(ctx : Text) → ""
+              , None =
+                  "cat " ++ Text/concatSep " " desc.dockerfilePaths ++ " | "
+              }
+              desc.dockerContext
+        ++  Text/concatSep
+              " "
+              ( List/map
+                  Text
+                  Text
+                  escapeShellArg
+                  ([ "docker" ] # mkArgs tag desc serviceDesc)
+              )
+
+let cloudBuild
+    : DockerfileDescription.Type →
+      ServiceDescription.Type →
+        Schema.Cloudbuild.Type
+    = λ(desc : DockerfileDescription.Type) →
+      λ(serviceDesc : ServiceDescription.Type) →
+        let tag = "gcr.io/\${PROJECT_ID}/${desc.service}:${serviceDesc.version}"
+
+        let script = mkScript tag desc serviceDesc
+
+        in  Schema.Cloudbuild::{
+            , steps =
+              [ Schema.Step::{
+                , name = "gcr.io/cloud-builders/docker"
```

Comment:
Are we sure we're actually using Kaniko, and caching the layers? Because if I read [this post](https://cloud.google.com/blog/products/containers-kubernetes/introducing-kaniko-build-container-images-in-kubernetes-and-google-container-builder-even-without-root-access) correctly, we need to use `gcr.io/kaniko-project/executor:latest` instead of `gcr.io/cloud-builders/docker`.

@balsoft Am I allowed to test out if this would actually cache the layers?

Reply:
This setup is not using kaniko's full caching capability, but I also think the staged dockerfile needs to be updated to make use of that cache consistently too when stages are identical